library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,64 +374,151 @@ impl<W: Write> BufWriter<W> {
+        // flushing, whereas `write` prefers not split incoming bufs.
+
+        // Bypass the buffer if the the incoming write is larger than the
+        // whole buffer. Use a vectored write to attempt to write the new
+        // data and the existing buffer in a single operation
+        let buf = match buf.len() >= self.capacity() {
+            true => match tail!(self.flush_buf_vectored(buf)?) {
+                // If the vectored write flushed everything at once, we're done!
+                [] => return Ok(()),
+
+                // If what's left after the vector flush is *still* larger than
+                // the buffer, bypass the buffer and forward it directly
+                tail if tail.len() >= self.capacity() => return self.get_mut().write_all(tail),
+
+                // Otherwise, we're going to buffer whatever's left of the user input
+                tail => tail,
+            },
+            false => buf,
+        };
+
+        // In order to reduce net writes in aggregate, we buffer as much as
+        // possible, then forward, then buffer the rest
+        let buf = tail!(self.write_to_buf(buf));
+        if !buf.is_empty() {
+            let buf = tail!(self.flush_buf_vectored(buf)?);
+
+            // At this point, because we know that buf.len() < self.buf.len(),
+            // and that the buffer has been flushed we know that this will
+            // succeed in totality
+            self.write_to_buf(buf);
         }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if buf.len() >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_all(buf);
-            self.panicked = false;
-            r
-        } else {
-            self.buf.extend_from_slice(buf);
-            Ok(())
+
+        // If, at this point, the buffer is full, we may as well eagerly
+        // attempt to flush, so that the next write will have an empty
+        // buffer.
+        if self.available() == 0 {
+            self.flush_buf()?;
         }
+
+        Ok(())
     }
 
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        let total_len = bufs.iter().map(|b| b.len()).sum::<usize>();
-        if self.buf.len() + total_len > self.buf.capacity() {
-            self.flush_buf()?;
-        }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if total_len >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_vectored(bufs);
-            self.panicked = false;
-            r
+        if let Some(buf) = only_one(bufs, |b| !b.is_empty()) {
+            // If there's exactly 1 incoming buffer, `Self::write` can make
+            // use of self.inner.write_vectored to attempt to combine flushing
+            // the existing buffer with writing the new one.
+            self.write(buf)
```

Comment:
I'm not sure this special case is worth optimizing for. `write_vectored` is used when there are multiple slices to write, and all the other paths handle occasionally empty slices correctly. The added branching here adds overhead to the most typical case of multiple non-empty slices.

Reply:
Hmm. I'll see if I can whip up a benchmark; my instinct is that branch prediction means that this overhead will be completely negligible (since, as you said, most of the time there'll be several inputs.

The major point of differentiation between `write` and `write_vectored` here doesn't have anything to do with empty slices; it's that `write` gets to use `flush_buf_vectored`.

Reply:
I went back and forth a lot on this, but I'm coming around to getting rid of it since, yeah, callers will likely end up calling the correct method. I've reviewed my *own* use of `write_vectored` in this PR (`flush_buf_vectored` etc) and confirmed that this usually ends up being the case. I'd love to get another opinion though (@m-ou-se?)